src/hotspot/share/runtime/mountUnmountDisabler.cpp

```diff
@@ -0,0 +1,430 @@
+      }
+      _current->rebind_to_jvmti_thread_state_of(_is_mount ? _vthread() : _current->threadObj());
+    }
+  }
+};
+
+class JVMTIEndTransition : public StackObj {
+  JavaThread* _current;
+  Handle _vthread;
+  bool _is_mount;
+  bool _is_thread_start;
+ public:
+  JVMTIEndTransition(JavaThread* current, oop vthread, bool is_mount, bool is_thread_start) :
+    _current(current), _vthread(current, vthread), _is_mount(is_mount), _is_thread_start(is_thread_start) {
+    if (DoJVMTIVirtualThreadTransitions && MountUnmountDisabler::notify_jvmti_events()) {
+      bool is_virtual = java_lang_VirtualThread::is_instance(_current->jvmti_vthread());
+      bool should_rebind = (_is_mount && !is_virtual) || (!_is_mount && is_virtual);
+      if (should_rebind) {
+        // We rebinded on start_transition but the mount/unmount operation
+        // failed so now we need to rebind to the original state.
+        _current->rebind_to_jvmti_thread_state_of(_is_mount ? _vthread() : _current->threadObj());
+        OrderAccess::fence();  // dont reorder with clear of transition flags
+      }
+    }
+  }
+  ~JVMTIEndTransition() {
+    if (DoJVMTIVirtualThreadTransitions && MountUnmountDisabler::notify_jvmti_events()) {
+      if (!_is_mount && _current->is_carrier_thread_suspended()) {
+        MonitorLocker ml(VTMSTransition_lock);
+        while (_current->is_carrier_thread_suspended()) {
+          ml.wait(200);
+        }
+      }
+
+      if (_is_thread_start) {
+        // If interp_only_mode has been enabled then we must eagerly create JvmtiThreadState
+        // objects for globally enabled virtual thread filtered events. Otherwise,
+        // it is an important optimization to create JvmtiThreadState objects lazily.
+        // This optimization is disabled when watchpoint capabilities are present. It is to
+        // work around a bug with virtual thread frames which can be not deoptimized in time.
+        if (JvmtiThreadState::seen_interp_only_mode() ||
+            JvmtiExport::should_post_field_access() ||
+            JvmtiExport::should_post_field_modification()){
+          JvmtiEventController::thread_started(_current);
+        }
+        if (JvmtiExport::should_post_vthread_start()) {
+          JvmtiExport::post_vthread_start((jthread)_vthread.raw_value());
+        }
+      }
+      if (_is_mount && JvmtiExport::should_post_vthread_mount()) {
+        JvmtiExport::post_vthread_mount((jthread)_vthread.raw_value());
+      }
+    }
+  }
+};
+#endif // INCLUDE_JVMTI
+
+bool MountUnmountDisabler::is_start_transition_disabled(JavaThread* thread, oop vthread) {
+  int base_disable_count = notify_jvmti_events() ? 1 : 0;
+  return java_lang_Thread::VTMS_transition_disable_count(vthread) > 0
+         || global_start_transition_disable_count() > base_disable_count
+         JVMTI_ONLY(|| (JvmtiVTSuspender::is_vthread_suspended(java_lang_Thread::thread_id(vthread)) || thread->is_suspended()));
+}
```

Comment:
I like this approach with the JVMTIStartTransition and JVMTIEndTransition helper classes. It is a nice way to decouple the JVMTI part of the protocol. Introducing the `is_start_transition_disabled()` function was also long desired.

Reply:
This is the function that needs a comment why you're testing all these things (and why base_disable_count is one for JVMTI). It's nice as a function that tests all the different values.